python_modules/dagster/dagster/_core/definitions/unresolved_asset_sensor_definition.py

```diff
@@ -0,0 +1,279 @@
+import json
+from collections import defaultdict
+from typing import TYPE_CHECKING, Mapping, Optional, Sequence, Set, Tuple
+
+import toposort
+
+from dagster._annotations import experimental
+
+from .asset_selection import AssetSelection
+from .events import AssetKey
+from .run_request import RunRequest
+from .sensor_definition import DefaultSensorStatus, SensorDefinition
+from .utils import check_valid_name
+
+if TYPE_CHECKING:
+    from dagster._core.definitions import AssetsDefinition, SourceAsset
+    from dagster._core.storage.event_log.base import EventLogRecord
+
+
+@experimental
+class UnresolvedAssetSensorDefinition:
+    def __init__(
+        self,
+        asset_selection: AssetSelection,
+        name: str,
+        minimum_interval_seconds: Optional[int] = None,
+        description: Optional[str] = None,
+        default_status: DefaultSensorStatus = DefaultSensorStatus.STOPPED,
+    ):
+        # TODO - type checks
+        self._selection = asset_selection
+        self._name = check_valid_name(name)
+        self._minimum_interval_seconds = minimum_interval_seconds
+        self._description = description
+        self._default_status = default_status
+
+        self._asset_keys = None
+        self._source_asset_keys = None
+
+    @property
+    def name(self):
+        return self._name
+
+    def _get_upstream_mapping(
+        self,
+        assets: Sequence["AssetsDefinition"],
+        source_assets: Sequence["SourceAsset"],
+    ) -> Mapping[AssetKey, Set[AssetKey]]:
+        """Computes a mapping of assets in self._selection to their parents in the asset graph"""
+        upstream = defaultdict(set)
+        selection_resolved = list(self._selection.resolve([*assets, *source_assets]))
+        for a in selection_resolved:
+            a_parents = list(AssetSelection.keys(a).upstream().resolve([*assets, *source_assets]))
```

Comment:
It's possible if we had assets A -> B -> C and we only selected asset C that this would fetch all upstreams (including both A and B) right?

Should we only be selecting `upstream(depth=1)`? As in only the immediate upstream parents of the selected assets

Reply:
good catch! i misread the docstring on `upstream`